packages/contracts-bedrock/src/L1/DataAvailabilityChallenge.sol

```diff
@@ -0,0 +1,343 @@
+        if (_challenge.resolvedBlock != 0) return ChallengeStatus.Resolved;
+
+        // if the challenge's start block is in the resolve window, it is active
+        if (_isInResolveWindow(_challenge.startBlock)) return ChallengeStatus.Active;
+
+        // if the challenge's start block is not in the resolve window, it is expired
+        return ChallengeStatus.Expired;
+    }
+
+    /// @notice Challenge a data commitment at a given block number.
+    /// @dev The block number parameter is necessary for the contract to verify the challenge window,
+    ///      since the contract cannot access the block number of the commitment.
+    ///      The function reverts if the caller does not have a bond or if the challenge already exists.
+    /// @param challengedBlockNumber The block number at which the commitment was made.
+    /// @param challengedHash The data commitment that is being challenged.
+    function challenge(uint256 challengedBlockNumber, bytes32 challengedHash) external payable {
+        // deposit value sent with the transaction as bond
+        deposit();
+
+        // require the caller to have a bond
+        if (balances[msg.sender] < bondSize) {
+            revert BondTooLow(balances[msg.sender], bondSize);
+        }
+
+        // require the challenge status to be uninitialized
+        if (getChallengeStatus(challengedBlockNumber, challengedHash) != ChallengeStatus.Uninitialized) {
+            revert ChallengeExists();
+        }
+
+        // require the current block to be in the challenge window
+        if (!_isInChallengeWindow(challengedBlockNumber)) {
+            revert ChallengeWindowNotOpen();
+        }
+
+        // reduce the caller's balance
+        balances[msg.sender] -= bondSize;
+
+        // store the challenger's address, bond size, and start block of the challenge
+        challenges[challengedBlockNumber][challengedHash] =
+            Challenge({ challenger: msg.sender, lockedBond: bondSize, startBlock: block.number, resolvedBlock: 0 });
+
+        // emit an event to notify that the challenge status is now active
+        emit ChallengeStatusChanged(challengedHash, challengedBlockNumber, ChallengeStatus.Active);
+    }
+
+    /// @notice Resolve a challenge by providing the pre-image data of the challenged commitment.
+    /// @dev The provided pre-image data is hashed (keccak256) to verify that it matches the challenged commitment.
+    ///      The function reverts if the challenge is not active or if the resolve window is not open.
+    /// @param challengedBlockNumber The block number at which the commitment was made.
+    /// @param preImage The pre-image data corresponding to the challenged commitment.
+    function resolve(uint256 challengedBlockNumber, bytes32 challengedHash, bytes calldata preImage) external {
+        // require the provided input data to match the commitment
+        if (challengedHash != keccak256(preImage)) {
```

Comment:
Why is the challenged hash in the calldata?

Reply:
Does it make the client implementation that much easier? Seems like you could compute the hash client side to be able to differentiate on the error kinds you mentioned above?

Reply:
Not _that much easier_, just a small ux improvement. I think this would become more relevant once there are different commitment types and the client would have to duplicate all the logic to compute the commitment from the pre-image, but with `keccak256` it's easy enough, so keeping the `challengedHash` param is not a strong opinion on my side.

Reply:
Another way to think about this is a challenge is keyed on (block number, input hash). We provide the full key here as args.

We could technically remove the input hash and derive the key by hashing the input data, but is less clear and requires duplicating hashing logic to do the same validation in the frontend.

Is there a strong reason to remove this?

Reply:
Just added versioned commitments (as discussed on slack). With that there is another reason to keep the `challengedCommitment` parameter, since it includes the commitment type based on which the resolve logic is branched.